src/serializer/Emitter.js

```diff
@@ -17,64 +17,119 @@ import { BodyReference } from "./types.js";
+//    To this end, the emitter maintains the `_activeBodies` and `_waitingForBodies` datastructures.
 export class Emitter {
   constructor(residualFunctions: ResidualFunctions) {
+    let mainBody = [];
     this._waitingForValues = new Map();
-    this._body = [];
+    this._waitingForBodies = new Map();
+    this._body = mainBody;
     this._declaredAbstractValues = new Set();
     this._residualFunctions = residualFunctions;
     this._activeStack = [];
     this._activeValues = new Set();
+    this._activeBodies = new Set([mainBody]);
+    this._finalized = false;
   }
 
+  _finalized: boolean;
   _activeStack: Array<string | Generator | Value>;
   _activeValues: Set<Value>;
+  _activeBodies: Set<Array<BabelNodeStatement>>;
   _residualFunctions: ResidualFunctions;
   _waitingForValues: Map<
     Value,
     Array<{ body: Array<BabelNodeStatement>, dependencies: Array<Value>, func: () => void }>
   >;
+  _waitingForBodies: Map<Array<BabelNodeStatement>, Array<{ dependencies: Array<Value>, func: () => void }>>;
   _declaredAbstractValues: Set<AbstractValue>;
   _body: Array<BabelNodeStatement>;
 
   beginEmitting(dependency: string | Generator | Value, targetBody: Array<BabelNodeStatement>) {
+    invariant(!this._finalized);
     this._activeStack.push(dependency);
     if (dependency instanceof Value) this._activeValues.add(dependency);
+    else if (dependency instanceof Generator) this._activeBodies.add(targetBody);
     let oldBody = this._body;
     this._body = targetBody;
     return oldBody;
   }
   emit(statement: BabelNodeStatement) {
+    invariant(!this._finalized);
     this._body.push(statement);
+    this._processCurrentBody();
   }
   endEmitting(dependency: string | Generator | Value, oldBody: Array<BabelNodeStatement>) {
+    invariant(!this._finalized);
     let lastDependency = this._activeStack.pop();
     invariant(dependency === lastDependency);
     if (dependency instanceof Value) {
       invariant(this._activeValues.has(dependency));
       this._activeValues.delete(dependency);
-      this._process(dependency);
+      this._processValue(dependency);
+    } else if (dependency instanceof Generator) {
+      invariant(this._activeBodies.has(this._body));
+      this._activeBodies.delete(this._body);
     }
     let lastBody = this._body;
     this._body = oldBody;
     return lastBody;
   }
-
-  _process(value: Value) {
+  finalize() {
+    invariant(!this._finalized);
+    invariant(this._activeBodies.size === 1);
+    invariant(this._activeBodies.has(this._body));
+    this._activeBodies.delete(this._body);
+    this._processCurrentBody();
+    this._finalized = true;
+    invariant(this._waitingForBodies.size === 0);
+    invariant(this._waitingForValues.size === 0);
+    invariant(this._activeStack.length === 0);
+    invariant(this._activeValues.size === 0);
+    invariant(this._activeBodies.size === 0);
+  }
+  _processCurrentBody() {
+    let a = this._waitingForBodies.get(this._body);
```

Comment:
I pondered this for far too long. It seems that somehow the actual body being processed is kept inside this._waitingForBodies, and that its representation is a list of [dependencies, emitterFunc] tuples. This is far from obvious and even now I'm not sure I understand what is going on here.